projects/npm-tools/packages/npm-scripts/contrib/prettier/prettier.js

```diff
@@ -1080,11 +1107,23 @@ function prepare(filepath) {
 
 	let file = 'prettier';
 
-	let root = getPortalRoot(filepath);
+	const workspaceRoot = getWorkspaceRoot(filepath);
+	const portalRoot = getPortalRoot(filepath);
 
-	if (root) {
-		const modules = path.join(root, 'modules/node_modules');
+	const portalOrWorkspaceRoot =
+		getPortalRoot(filepath) || getWorkspaceRoot(filepath);
```

Comment:
We are adding some unnecessary work her, we should just crawl the directories once

```js
	const workspaceRoot = getWorkspaceRoot(filepath);
	const portalRoot = getPortalRoot(filepath);

    const portalOrWorkspaceRoot = portalRoot || workspaceRoot;
```

Even better, we may consider creating a generic `findRoot` function that only has to crawl the file structure once and then returns what _type_ of root it is, whether it be a workspace or liferay-portal. This would allow us to avoid spending extra resources on crawling once for workspace and once for portal

Reply:
This is a good suggestion. Here's what I ended up with, it's faulty and untested:
```js
function findRoot(filepath){
	let current = filepath;

	while (true) {
	const parent = path.dirname(current);

	const portalRootCandidate = path.join(parent, 'portal-web');
	const frontendProjectsRootCandidate = path.join(parent, '.liferay-frontend-projects');
	const workspaceRootCandidate = path.join(parent, 'settings.gradle');

	const candidates = [portalRootCandidate, frontendProjectsRootCandidate, workspaceRootCandidate];

	const candidate = candidates.find(candidate => {
		if (candidate === portalRootCandidate) {
			return fs.existsSync(
				path.join(current, 'docroot/WEB-INF/liferay-web.xml')
			) ? true : false;
		}

		if (candidate === workspaceRootCandidate) {
			return fs.existsSync(candidate).includes(
				'apply plugin: "com.liferay.workspace"'
			) ? true : false;
		}

		return fs.existsSync(candidate);
	})

		if (candidate) {
			current = candidate;

			break;
		}
		else if (parent === current) {

			// Can't go any higher.

			return null;
		}
		else {
			current = parent;
		}
	}
}
```

I wonder how will the `prepare` function be able to tell what root it is, like we are currently doing:
```
const workspaceRoot = getWorkspaceRoot(filepath);

if (workspaceRoot) {
...
}
```

There are definitely improvements to the `findRoot` function that would make it possible, but I just don't see it right now. It's gonna require a bit more time than what I have today, so I won't push anything since it won't do anything good.